### PR TITLE
[new] Advanced evoting DApp Avalanche

### PR DESCRIPTION
**Pull request by rajranjan0608**
# Description*

We will be building a more advanced e-voting `dApplication`, in which we will not only interact with pre-deployed smart contracts, but will also deploy the contracts in runtime, and would interact with them using their contract address. For developing this dApplication we would be using the `Trufflesuite` framework.

## Type of change*

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## This PR created for* - 
- [x] DataHub Official Docs (Network Documentation, API docs etc.) changes/updates/fixes
- [x] Figment Learn Docs (Pathway, Tutorials, New Tutorials etc.) changes/updates/fixes

# Checklist*:

- [x] I updated the `SUMMARY.md` file (FYI, It's needed only when you're creating a new file like submitting a new tutorial, pathway section)
- [x] I created one tutorial for this PR (one PR <=> one tutorial)
- [x] I checked and read the contribution guide before making this PR [**HERE**](https://docs.google.com/document/d/13LWLrWzZ34M0ldWGeDANcWxw9nEWk3AX3VwXRBIOs1M/edit)  
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by yash-sharma1** [comment]
LGTM 👌 @pyr0gan I need some help 👀 & in running it locally on the machine, currently, I'm getting some errors in installing some dependencies, can you please test it on your local and after that, we can merge it.

**Comment by rajranjan0608** [comment]
> LGTM  @pyr0gan I need some help  & in running it locally on the machine, currently, I'm getting some errors in installing some dependencies, can you please test it on your local and after that, we can merge it.

Hey, @yash-sharma1 Please let me know the error, so that, I can give a quick fix.

**Comment by yash-sharma1** [comment]
> > LGTM  @pyr0gan I need some help  & in running it locally on the machine, currently, I'm getting some errors in installing some dependencies, can you please test it on your local and after that, we can merge it.
> 
> Hey, @yash-sharma1 Please let me know the error, so that, I can give a quick fix.

Thanks for the concern @rajranjan0608 but seems like it does not cause of you, might be it's my machine, throwing `ERR 128` even after having everything setup and configured with github.

**Comment by yash-sharma1** [comment]
> @yash-sharma1 cool then. You might want to try these fixes
> 
> * Try updating npm - `sudo npm install -g npm`
> * Use `sudo` in the commands wherever we are installing `npm packages` globally, like in `sudo npm install -g truffle`.

Thanks Raj Ranjan, It was actually due to my machine and some broken modules, just got it fixed :)

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 1]
```diff
@@ -0,0 +1,855 @@
+---
+description: Learn how to use Truffle with the C-Chain
+---
+
+# Making an advanced e-Voting dApp on Avalanche Fuji network using Trufflesuite
+
+## About the author
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:7]: Can you move this at the very end of the tutorial?

cc @yash-sharma1 let's ask this going forward as we discussed

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+
+Hi, I am [Raj Ranjan](https://rajranjan0608.github.io) and currently I am doing my graduation \(B.Tech, 2018-22\) in Computer Science from IIIT Guwahati, India. I love building blockchain applications and is always looking for oppurtunities to give back to the community. You may reach out to me on [Linkedin](https://www.linkedin.com/in/iamrajranjan), if you want to discuss about any project ideas, have doubt or anything like that. Hope you would like the tutorial ahead :\)
+
+## Introduction
+
+Hey, everyone. In the last tutorial on [Making a basic e-voting DApp on Avalanche's Fuji C-Chain](https://learn.figment.io/network-documentation/avalanche/tutorials/making-evoting-dapp-on-avalanche-c-chain-using-truffle), we have learnt how to deploy a smart contract on Avalanche's Fuji C-Chain using Trufflesuite. Along with that we have also coded the client-side application, to easily interact with smart contracts.
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:13]: can you remove the conversational/oral language `Hey, everyone`?

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 3]
```diff
@@ -0,0 +1,855 @@
...
+
+Today, we will be building a more advanced e-voting `dApplication`, in which we will not only interact with pre-deployed smart contracts, but will also deploy the contracts in runtime, and would interact with them using their contract address. For developing this dApplication we would be using Trufflesuite framework. 
+
+For your information, [Truffle Suite](https://www.trufflesuite.com) is a toolkit for launching decentralized applications \(dapps\) on the EVM. With Truffle you can write and compile smart contracts, build artifacts, run migrations and interact with deployed contracts. This tutorial illustrates how Truffle can be used with Avalanche's C-Chain, which is an instance of the EVM.
+
+## Requirements
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:19]: You've switched "Requirements" and "Prerequisites".

- Requirements => node, packages, etc.
- Prerequisites => useful/recommended prior knowledge

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 4]
```diff
@@ -0,0 +1,855 @@
...
+Create and enter a new directory named `advanced-evoting`:
+
+```javascript
+mkdir advanced-evoting && cd advanced-evoting
+```
+
+Use `npm` to install other dependencies
+
+```text
+npm init
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:49]: i did this and it asked me for prompts but this tutorial is not telling me what to do with the prompts
Two ways of fixing this:
1) explain to the user that they can just breeze through the prompts by hitting Enter
2) `npm init -y` which will skip the prompts.

In other tutorials we do 2 so i'd recommend doing the same

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+
+  //Public vote function for voting a candidate
+  function vote (uint _candidate) public {
+    require(!voters[msg.sender], "Voter has already Voted!");
+    require(_candidate < candidatesCount && _candidate >= 0, "Invalid candidate to Vote!");
+    voters[msg.sender] = true;
+    candidates[_candidate].voteCount++;
+  }
+}
+```
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:167]: Can you explain what this contract does, line block by line block?

We need to do more than just ask learners to copy and paste pieces of code they don't understand

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+contract MainContract {
+    uint public electionId = 0;
+    mapping (uint => address) public Elections;
+
+    function createElection (string[] memory _nda, string[] memory _candidates) public {
+        Election election = new Election(_nda, _candidates);
+        Elections[electionId] = address(election);
+        electionId++;
+    }
+}
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:190]: same here

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+> Compiling ./contracts/Election.sol
+> Compiling ./contracts/MainContract.sol
+> Compiling ./contracts/Migrations.sol
+
+> Artifacts written to /home/guest/blockchain/advanced-evoting/build/contracts
+> Compiled successfully using:
+   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang
+```
+
+Compiling the smart contracts would create `.json` file in the `build/contracts` directory. It stores `ABI` and other necessary metadata.
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:231]: What are `ABI`?

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 8]
```diff
@@ -0,0 +1,855 @@
...
+
+## Fund the account and run migrations on the C-Chain
+
+When deploying smart contracts to the C-Chain, it will require some deployment cost. As you can see inside `truffle-config.js`, HDWallet Provider will help us in deploying on Fuji C-chain and deployment cost will be managed by account whose mnemonic has been stored in the `.env` file. Therefore we need to fund the account.
+
+### Fund your account
+
+Fund your account using the the faucet link [https://faucet.avax-test.network/](https://faucet.avax-test.network/) and pasting your Fuji's C-Chain address in the input field or follow the steps in the [Transfer AVAX Between X-Chain and C-Chain](https://learn.figment.io/network-documentation/avalanche/tutorials/transfer-avax-between-the-x-chain-and-c-chain) tutorial to fund the newly created account. You'll need to send at least `135422040` nAVAX to the account to cover the cost of contract deployments.
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:239]: this says to paste the C-Chain address and then follow a tutorial to transfer from X-chain to C-Chain.

Why do i need to transfer from X to C if i entered my C address in the faucet?

**Review by guillaumegaluz — CHANGES_REQUESTED** [review]
Hey @rajranjan0608 I did a first pass at this

**Comment by guillaumegaluz** [comment]
Screen Shot 2021-05-10 at 4 50 33 PM

Is this changing the "DataHub Official Docs"?

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md (hunk 1 above)
Comment (yash-sharma1) [reply network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:7]: Sure Thing
@rajranjan0608 can you shift the "About the Author" section at the end of the tutorial I mean after the Conclusion

Comment (rajranjan0608) [reply network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:7]: Yeah sure.

Comment (yash-sharma1) [reply network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:7]: Also, make the same change in your previous e-voting DApp tutorial to make it consistent.

Comment (rajranjan0608) [reply network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:7]: OK. Actually, I'm working on it.

**Comment by yash-sharma1** [comment]
> @pyr0gan @yash-sharma1 can you both run this from scratch as if you were a learner and confirm you can go throught the whole tutorial?

Yes, I was going through it and there an error related to some truffle/fuji upgrades for which a couple of members and @rajranjan0608 are trying to fix and that's why we're keeping it on hold. Till the step of the truffle requirement, it's working fine.

File: network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md [hunk 9]
```diff
@@ -43,10 +39,16 @@ Create and enter a new directory named `advanced-evoting`:
 mkdir advanced-evoting && cd advanced-evoting
 ```
 
-Use `npm` to install other dependencies
+First initialize the folder with `npm` for, in order to have organise the project.
 
 ```text
 npm init
```
Comment (yash-sharma1) [inline network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:45]: @rajranjan0608 it should be `npm init -y`

Comment (rajranjan0608) [reply network-documentation/avalanche/tutorials/making-advanced-e-voting-dapp-avalanche-fuji-using-trufle.md:45]: It could be an option to use `-y` if we want to skip all prompts, but I have mentioned about manually skipping a prompt one by one by hitting enter.

**Comment by guillaumegaluz** [comment]
The other thing we should change is write the FE in React instead if jQuery. 
Would also help differentiate the tutorial from https://github.com/dappuniversity/election which it is copied from

**Comment by rajranjan0608** [comment]
> The other thing we should change is write the FE in React instead if jQuery.
> Would also help differentiate the tutorial from https://github.com/dappuniversity/election which it is copied from

Yes, there could be a separate tutorial for building the frontend in ReactJS. Also, I have taken inspiration from Dapp University and made this tutorial. The following could help in differentiating between the two -
1. Added deployable Election contracts by using MainContract.sol
2. Integrated javascript code to interact with Election contracts using their address.
3. Organised the handling of different Election contracts in the frontend, by implementing their class (`./src/Election.js`)

**Comment by guillaumegaluz** [comment]
> > The other thing we should change is write the FE in React instead if jQuery.
> > Would also help differentiate the tutorial from https://github.com/dappuniversity/election which it is copied from
> 
> Yes @guillaumegaluz, there could be a separate tutorial for building the frontend in ReactJS. But this tutorial with its frontend in regular javascript is necessary because it deals with a different concept of deploying instances of 1 smart contract (`Election.sol`) from another contract (`MainContract.sol`) and interacting with it using their address. This version of the tutorial is not available at Dapp University.
> 
> I have taken inspiration from Dapp University and made this tutorial. The following could help in differentiating between the two -
> 
> 1. Added deployable Election contracts by using MainContract.sol
> 2. Integrated javascript code to interact with Election contracts using their address.
> 3. Organised the handling of different Election contracts in the frontend, by implementing their class (`./src/Election.js`)

where is the JS code coming from? Nobody really writes jQuery anymore.

**Comment by rajranjan0608** [comment]
> where is the JS code coming from? Nobody really writes jQuery anymore.

@guillaumegaluz Those who don't know about ReactJS or just started with basic HTML, CSS and (vanilla) javascript and later moved to blockchain or DApp development, can be benefitted from this tutorial. We may work on separate tutorials for ReactJS.

**Comment by yash-sharma1** [comment]
> > where is the JS code coming from? Nobody really writes jQuery anymore.
> 
> @guillaumegaluz Those who don't know about ReactJS or just started with basic HTML, CSS and (vanilla) javascript and later moved to blockchain or DApp development, can be benefitted from this tutorial. We may work on separate tutorials for ReactJS.

Hey @rajranjan0608 
We were working on few things to make tutorials more consistent and well structured because of that your PR is still under review. So moving further I request you please follow this guide **[HERE](https://learn.figment.io/other/tutorial-guidelines)** to make some required & requested changes in your PR (both new and previous tutorial). Let us know if you have any queries.

**Comment by rajranjan0608** [comment]
Hi @yash-sharma1. I have made changes as per the new tutorial guidelines in the previous as well as the current tutorial. Please have a look. Also, I have committed the previous tutorial updates in the same branch and can be referenced from here - 4b730a592588a68c0ecfc8c50c6f33f1898e0010

File: network-documentation/avalanche/tutorials/making-evoting-dapp-on-avalanche-c-chain-using-truffle.md [hunk 10]
```diff
@@ -60,7 +63,7 @@ One of the files created when you ran `truffle init` is `truffle-config.js`. Rep
 
 ```javascript
 require('dotenv').config();
-const HDWalletProvider = require("truffle-hdwallet-provider");
+const HDWalletProvider = require("@truffle/hdwallet-provider");
```
Comment (guillaumegaluz) [inline network-documentation/avalanche/tutorials/making-evoting-dapp-on-avalanche-c-chain-using-truffle.md:67]: since we changed the package from `truffle-hdwallet-provider` to `@truffle/hdwallet-provider` we also need to use the new interface.

```
let provider = new HDWalletProvider(mnemonic, ...)
```

is deperacted. The new interface is documented [here](https://github.com/trufflesuite/truffle/tree/master/packages/hdwallet-provider#truffle-usage)

So the config needs to change from:

```
module.exports = {
  networks: {
    fuji: {
      provider: function() {
        return new HDWalletProvider(mnemonic, `https://avalanche--fuji--rpc.datahub.figment.io/apikey/${APIKEY}/ext/bc/C/rpc`)
      },
      ...
    }
  },
  ...
}
```

to

```
module.exports = {
  networks: {
    fuji: {
      provider: function() {
        return new HDWalletProvider({
          mnemonic: {
            phrase: mnemonic
          },
          providerOrUrl: `https://avalanche--fuji--rpc.datahub.figment.io/apikey/${APIKEY}/ext/bc/C/rpc`,
        });
      },
      ...
    }
  },
  ...
}
```

I get an error in the terminal:

Screen Shot 2021-05-12 at 10 13 25 PM

and when i check the logs on Datahub i see:

Screen Shot 2021-05-12 at 10 11 30 PM

whcih is weird cause the status in `200`...

Bottom line: i dont think that as is the first tutorial is working. We forked it from Dapp University which is 2 year old and we need to adapt it to make it work with the new Truffle package

Comment (yash-sharma1) [reply network-documentation/avalanche/tutorials/making-evoting-dapp-on-avalanche-c-chain-using-truffle.md:67]: Yes, Guillaume you're right. I run this on local last night before signing off and I had the same 'pify' error. Pinged the Raj Ranjan, as he mentioned all worked for him but not for me.

Comment (rajranjan0608) [reply network-documentation/avalanche/tutorials/making-evoting-dapp-on-avalanche-c-chain-using-truffle.md:67]: @guillaumegaluz Yes, we should update the interface as you mentioned. Along with that, we have to add one more parameter to the object which is `chainId: "0xa869"`. This chain ID is given in Avalanche's official documentation [here](https://docs.avax.network/build/tutorials/smart-contracts/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask) It's strange because this chainId is not required when we are using the actual `RPC` URL of the Avalanche's Fuji network viz. `https://api.avax-test.network/ext/bc/C/rpc`. Might be Figment need to update their API according to the new package update. 
So, our new interface would look like this (I'll update this in the new commit ASAP) -

```javascript
return new HDWalletProvider({
   mnemonic: {
      phrase: mnemonic
   },
   providerOrUrl: `https://avalanche--fuji--rpc.datahub.figment.io/apikey/${APIKEY}/ext/bc/C/rpc`,
   chainId: "0xa869"
});
```
But I didn't get the error regarding `pify`.

Comment (yash-sharma1) [reply network-documentation/avalanche/tutorials/making-evoting-dapp-on-avalanche-c-chain-using-truffle.md:67]: Thanks, the error regarding `pify` is resolved but still not able to deploy the contract as Guillaume mentioned

**Comment by rajranjan0608** [comment]
@yash-sharma1 and @guillaumegaluz Please verify it once again. I have updated the new interface for HDWalletProvider as needed. Let me know if you are still getting some error.

**Comment by guillaumegaluz** [comment]
1) re-pify: Raj, you might have installed it globally which is why npm is not complaining? Yash and I bought had it so I would just add a note that says "if you're getting `Error: Cannot find module 'pify'`, run `npm install pify --save`"

2) re-`truffle migrate`: We've increased the # of concurrent requests that are allowed through DataHub (from 5 to 20). I ran `truffle migrate --network fuji` again and was able to deploy the contracts! [See screenshot](https://share.getcloudapp.com/2NuwL7j9). 

3) Taking a step back on those technical errors, can you summarize the difference between the existing evoting tutorial and the Advanced evoting tutorial?

**Comment by rajranjan0608** [comment]
> 1. re-pify: Raj, you might have installed it globally which is why npm is not complaining? Yash and I bought had it so I would just add a note that says "if you're getting `Error: Cannot find module 'pify'`, run `npm install pify --save`"
> 2. re-`truffle migrate`: We've increased the # of concurrent requests that are allowed through DataHub (from 5 to 20). I ran `truffle migrate --network fuji` again and was able to deploy the contracts! [See screenshot](https://share.getcloudapp.com/2NuwL7j9).
> 3. Taking a step back on those technical errors, can you summarize the difference between the existing evoting tutorial and the Advanced evoting tutorial?

1. re-pify: That would be better to keep a note, in the tutorial, about this tutorial. But, FYI, I have also tried to run this tutorial on an isolated AWS console, and it worked. But can't say whether it was pre-installed there or not.

2. re-`truffle-migrate`: That's great. This error was related to `eth-block-tracker` package, which tries to update block, until it's successful or till a certain timeout. It might end up making many requests concurrently to the API, and reaching the limit. So, I think limit of 20 would be better, if timeout is not achieved before that.

3. Referencing to the errors, errors would be same in both of the tutorials. A brief summary about extra functionalities of advanced e-voting -
     * We can give details like name and description to each election in advanced e-voting.
     * We can deploy custom elections as much as we want, from the UI itself.
     * We can add any number of candidates in each election.
